storage: don't send historical Raft log with snapshots

tbg opened the pull request:
Assume a leaseholder wants to send a (Raft or preemptive) snapshot to a
follower. Say the leaseholder's log ranges from 100 to 200, and we assume
that the size
(in bytes) of this log is 200mb. All of the log is successfully committed
and applied, and is thus reflected in the snapshot data.

Prior to this change, we would still send the 200mb of log entries along
with the snapshot, even though the snapshot itself already reflected them.

After this change, we won't send any log entries along with the snapshot,
as sanity would suggest we would.

We were unable to make this change because up until recently, the Raft
truncated state (which dictates the first log index) was replicated and
consistency checked; this was changed in #34660.

Somewhere down the road (19.2?) we should localize all the log truncation
decisions and simplify all this further. I suspect that in doing so we can
avoid tracking the size of the Raft log in the first place; all we really
need for this is some mechanism that makes sure that an "idle" replica
truncates its logs. With unreplicated truncation, this becomes cheap enough
to "just do".

Release note (bug fix): Remove historical log entries from Raft snapshots.
These log entries could lead to failed snapshots with a message such as:

    snapshot failed: aborting snapshot because raft log is too large
    (25911051 bytes after processing 7 of 37 entries)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35701)
<!-- Reviewable:end -->

tbg commented:
I need to look at this again with a clear head tomorrow, but wanted to push this out so that @andreimatei can do preliminary review.

tbg commented:
PS I saw recent clearrange/checks=false runs with hundreds of snapshots failed due to size, I had thought that we had fixed all the issues with the log size but apparently there are always more. Hence my renewed will to do this in 19.1.

This will need a cluster version.

tbg commented:
> This will need a cluster version.

I take that back, I did all the hard work in the PR unreplicating the truncated state.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35701)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/storage/replica_command.go, line 899 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-L_tTu6XD0a7WojJaSLR:-L_tTu6Y4t-5aysdJAHP:bhhe4n2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1124f827da5b53b30ecf67bf5f17dc5bab3e8e/pkg/storage/replica_command.go#L899)):*
> ```Go
> 
> 	if !usesReplicatedTruncatedState && snap.State.TruncatedState.Index < snap.State.RaftAppliedIndex {
> 		// If we're not using a legacy (replicated) truncated state, we can
> ```

s/we can/we
___
*[pkg/storage/replica_command.go, line 906 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-L_tUafc1bJXnHPbZYJr:-L_tUafc1bJXnHPbZYJs:b8o5eg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1124f827da5b53b30ecf67bf5f17dc5bab3e8e/pkg/storage/replica_command.go#L906)):*
> ```Go
> 		// again we're relying on some mechanisms (quota pool, Raft uncommitted
> 		// size limit) to make sure this doesn't grow without bounds.
> 		snap.State.TruncatedState = &roachpb.RaftTruncatedState{
> ```

So stupid question here so that you get real confidence in this review: how exactly does this patch actually cause the logs to not be included in the snapshot? Does anybody on the sender side look at this `snap.State.TruncatedState`? :)
___
*[pkg/storage/replica_raft.go, line 2338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-L_tWlms-4cAIKXnfz0e:-L_tWlms-4cAIKXnfz0f:b-12ax93) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1124f827da5b53b30ecf67bf5f17dc5bab3e8e/pkg/storage/replica_raft.go#L2338)):*
> ```Go
> 			rResult.State.TruncatedState = nil
> 			rResult.RaftLogDelta = 0
> 			// We received a truncation that doesn't apply to us, so we know that
> ```

I believe this is correct, but the return value for `handleTruncatedStateBelowRaft()` really needs to be documented. Try to read that function without a tome of context (as I am :P ).

So to make sure I'm with you here, we don't expect this branch to be commonly taken, do we? This is taken is the leaseholder had more logs than we did, and to attempt to truncate up to an index that's below any that we have in our logs. How exactly would that happen?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35701)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/storage/replica_command.go, line 899 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-L_tTu6XD0a7WojJaSLR:-L_wJ3Br5CKD5lhcO7pZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1124f827da5b53b30ecf67bf5f17dc5bab3e8e/pkg/storage/replica_command.go#L899)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/we can/we
</blockquote></details>

Done.
___
*[pkg/storage/replica_command.go, line 906 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-L_tUafc1bJXnHPbZYJr:-L_wJXQLDGUdreyDLwoO:b-f3lpro) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1124f827da5b53b30ecf67bf5f17dc5bab3e8e/pkg/storage/replica_command.go#L906)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

So stupid question here so that you get real confidence in this review: how exactly does this patch actually cause the logs to not be included in the snapshot? Does anybody on the sender side look at this `snap.State.TruncatedState`? :)
</blockquote></details>

Yep, here:

https://github.com/cockroachdb/cockroach/blob/2beec4e912fbf7daf8b0cf4bafe4d43bf4ad6147/pkg/storage/store_snapshot.go#L212-L216

I also added a comment to that effect.

I also just figured out that with this change we're effectively not sending any log entries at all, which is great! One less thing to worry about.
___
*[pkg/storage/replica_raft.go, line 2338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-L_tWlms-4cAIKXnfz0e:-L_x45e472QzA1U-pqzx:bfj5wb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1124f827da5b53b30ecf67bf5f17dc5bab3e8e/pkg/storage/replica_raft.go#L2338)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I believe this is correct, but the return value for `handleTruncatedStateBelowRaft()` really needs to be documented. Try to read that function without a tome of context (as I am :P ).

So to make sure I'm with you here, we don't expect this branch to be commonly taken, do we? This is taken is the leaseholder had more logs than we did, and to attempt to truncate up to an index that's below any that we have in our logs. How exactly would that happen?
</blockquote></details>

You're right that this needed more words. I added some on `handleTruncatedStateBelowRaft`, WDYT?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

gold Jerry, gold

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-:-La0y1r04us59qh_QC8e:bwhqbup)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replica_command.go, line 905 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-La0xQvuDIR5lxvzHG8I:-La0xQvv5ZupqouvQN-7:b-nrjpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef8a2a4206a2cb74c3a50b99e5d2feb46354a62b/pkg/storage/replica_command.go#L905)):*
> ```Go
> 		// to do since it is not a replicated key (and thus not subject to
> 		// matching across replicas). The actual sending happens here:
> 		_ = (*kvBatchSnapshotStrategy)(nil).Send
> ```

lool I appreciate the innovative thinking, but this line is a bit much
___
*[pkg/storage/replica_raft.go, line 2338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-L_tWlms-4cAIKXnfz0e:-La0xucDCIVQVsUKejx9:b-7j76d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1124f827da5b53b30ecf67bf5f17dc5bab3e8e/pkg/storage/replica_raft.go#L2338)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You're right that this needed more words. I added some on `handleTruncatedStateBelowRaft`, WDYT?
</blockquote></details>

sweet


<!-- Sent from Reviewable.io -->

tbg commented:
TFTR! Going to run a few relevant roachtests off of this branch before I merge.

bdarnell reviewed (approved):
> Reviewed 4 of 6 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35701)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replica_raft.go, line 2349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-La6vL0A43hpZsQhPV11:-La6vL0A43hpZsQhPV12:b-k20ek1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8303ba074c8a6c7cd8698b55da2cb47f0934cde/pkg/storage/replica_raft.go#L2349)):*
> ```Go
> 			// Morally we would want to drop the command in checkForcedErrLocked,
> 			// but that may be difficult to achieve.
> 			log.Fatal(ctx, log.Safe(fmt.Sprintf("TruncatedState regressed:\nold: %+v\nnew: %+v", oldTruncatedState, rResult.State.TruncatedState)))
> ```

I was just about to add a test that relied on this assertion (when a TruncateLog operation applied twice). I guess I'll have to find a different failure signal for this test, but the strictness here was useful to bring a problem in #35261 to light. 
___
*[pkg/storage/replica_raftstorage.go, line 1002 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35701#-La6w46w2tmHbMFPAMvD:-La6w46w2tmHbMFPAMvE:b2pvhhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef8a2a4206a2cb74c3a50b99e5d2feb46354a62b/pkg/storage/replica_raftstorage.go#L1002)):*
> ```Go
> 	// wholesale replace r.mu.state.
> 	r.mu.state = s
> 	// Snapshots typically have less log entries than the leaseholder. The next
> ```

s/less/fewer/


<!-- Sent from Reviewable.io -->

bdarnell commented:
For release notes @jseldess: This change is unavailable until the upgrade to 19.1 is finalized (the cluster version was introduced in a separate PR that didn't have a release note: the "unreplicated truncated state" change)

tbg commented:
Kicked off runs for `clearrange/checks=false` and `import/tpcc/warehouses=1000/nodes=32`. They both run larger imports that I've recently found to cause issues with the large Raft log and that I would now expect to highlight problems related to this change. The former also runs lots of merges which aggressively move replicas for good coverage. Will report back in a few hours.

tbg committed:
storage: don't send historical Raft log with snapshots

Assume a leaseholder wants to send a (Raft or preemptive) snapshot to a follower.
Say the leaseholder's log ranges from 100 to 200, and we assume that the size
(in bytes) of this log is 200mb. All of the log is successfully
committed and applied, and is thus reflected in the snapshot data.

Prior to this change, we would still send the 200mb of log entries along
with the snapshot, even though the snapshot itself already reflected
them.

After this change, we won't send any log entries along with the
snapshot, as sanity would suggest we would.

We were unable to make this change because up until recently, the Raft
truncated state (which dictates the first log index) was replicated and
consistency checked; this was changed in #34660. The migration
introduced there makes it straightforward to omit a prefix of the
log in snapshots, as done in this commit.

Somewhere down the road (19.2?) we should localize all the log
truncation decisions and simplify all this further. I suspect
that in doing so we can avoid tracking the size of the Raft log
in the first place; all we really need for this is some mechanism
that makes sure that an "idle" replica truncates its logs. With
unreplicated truncation, this becomes cheap enough to "just do".

Release note (bug fix): Remove historical log entries from Raft snapshots.
These log entries could lead to failed snapshots with a message such as:

    snapshot failed: aborting snapshot because raft log is too large
    (25911051 bytes after processing 7 of 37 entries)

tbg commented:
`import/tpcc/warehouses=1000/nodes=32` finished like a charm in 13 minutes (including setting u the machines, etc). Looking at the test history, either this was a fairly lucky run or we've actually shaved a few minutes off the run with this change (16-18 minutes is more typical for this test).

jseldess commented:
Thanks for the ping, @bdarnell. I already have it in my list of "performance improvements" [here](https://docs.google.com/document/d/1uu0AOqbfGt5w5ORHWXZVW-FNFSWYiyJSzb9JAb1LE14/edit#heading=h.n07g1fqufz9k), though I don't think I'll specifically explain most of those.

tbg commented:
Ran another `import/tpcc/warehouses=1000/nodes=32` in 13 minutes end-to-end. `clearrange` is also looking good.
TFTRs!

bors r=andreimatei,bdarnell
